### PR TITLE
10: Test Case 4: Patient Search feature

### PR DESCRIPTION
**Pull request by sherrif10**
https://github.com/IsantePlus/isanteplus-qaframework/issues/10  cc @moshonk  @mozzy11 

Searching For ST CODE is the same as searching for a patient using first Name or Last Name so am thinking we can go with findPatient by name.

### CODE REVIEW

File: src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/test/TestBase.java [hunk 1]
```diff
@@ -195,9 +216,70 @@ protected void quit() {
 		}
 	}
 	
+	protected WebElement getElement(By elementBy) {
+		try {
+			return driver.findElement(elementBy);
+		} catch (NoSuchElementException e) {
+			return null;
+		}
+	}
```
Comment (mozzy11) [inline src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/test/TestBase.java:225]: This util already exists in the parent Page class.
we are even not supposed to use this anywhere in the step definitions.
This is supposed to be wrapped up in the Page objetcs

File: src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/test/TestBase.java [hunk 2]
```diff
@@ -195,9 +216,70 @@ protected void quit() {
...
+		assertTrue(ids.contains(trimPatientId(patientId)));
+	}
+	
 	public String getCurrentDate() {
 		Date date = new Date();
 		SimpleDateFormat formatter = new SimpleDateFormat("dd/MM/yyyy");
 		return formatter.format(date);
 	}
+	
+	protected void initiateWithLogin() {
```
Comment (mozzy11) [inline src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/test/TestBase.java:252]: we dont need this util , the function already exist in the Login page [here](https://github.com/IsantePlus/isanteplus-qaframework/blob/main/src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/page/LoginPage.java#L71-L78)

File: src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/test/TestBase.java (hunk 1 above)
Comment (sherrif10) [reply src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/test/TestBase.java]: Resolved thanks

File: src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/test/TestBase.java (hunk 2 above)
Comment (sherrif10) [reply src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/test/TestBase.java]: Resolved thanks

File: src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/FindPatientSteps.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	private ClinicianFacingPatientDashboardPage dashboardPage;
+	
+	private LoginPage loginPage;
+	
+	private HomePage homePage;
+	
+	@Before(RunTest.HOOK.FINDPATIENT)
+	public void systemLogin() {
+		homePage = loginPage.goToHomePage();
+
```
Comment (mozzy11) [inline src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/FindPatientSteps.java:32]: remove unnecesary space

File: src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/FindPatientSteps.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+	public void destroy() {
+		quit();
+	}
+	
+	@Given("User clicks on Find Patient App")
+	public void visitFindPatientPage() {
+		if (findPatientPage != null) {
+			findPatientPage = (FindPatientPage) homePage.goToFindPatientRecord().waitForPage();
+		}
+
```
Comment (mozzy11) [inline src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/FindPatientSteps.java:45]: remove unnecesary space

File: src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/page/ClinicianFacingPatientDashboardPage.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+	private static final By VIRAL_LOAD_HISTORY = By .id("isanteplus.viralLoadHistory");
+	private static final By PATIENT_SUMMARY = By.id("isanteplusreports.patientSummary");
+	private static final By DIAGNOSIS_HISTORY = By.id("isanteplus.diagnosis.label");
+	private static final By END_VISIT = By.id("referenceapplication.realTime.endVisit");
+
+
+
+	public ClinicianFacingPatientDashboardPage(Page parent) {
+		super(parent);
+
```
Comment (mozzy11) [inline src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/page/ClinicianFacingPatientDashboardPage.java:31]: remove unnecesary space

File: src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/page/LoginPage.java [hunk 6]
```diff
@@ -80,4 +75,10 @@ public HomePage goToHomePage() {
     public Boolean hasLoginButton() {
         return hasElement(BUTTON_LOGIN);
     }
+
+	public Page login(String user, String password) {
+		 String value = findElement(By.cssSelector("#sessionLocation li")).getAttribute("value");
+	        return login(user, password);
+
```
Comment (mozzy11) [inline src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/page/LoginPage.java:82]: remove unnecesary space

File: src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/page/LoginPage.java [hunk 7]
```diff
@@ -80,4 +75,10 @@ public HomePage goToHomePage() {
     public Boolean hasLoginButton() {
         return hasElement(BUTTON_LOGIN);
     }
+
+	public Page login(String user, String password) {
```
Comment (mozzy11) [inline src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/page/LoginPage.java:79]: What does this method do ??

File: src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/FindPatientSteps.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+	}
+	
+	@After(RunTest.HOOK.FINDPATIENT)
+	public void destroy() {
+		quit();
+	}
+	
+	@Given("User clicks on Find Patient App")
+	public void visitFindPatientPage() {
+		if (findPatientPage != null) {
```
Comment (mozzy11) [inline src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/FindPatientSteps.java:42]: @sherrif10 , what is this here ?????
is there any situation where  findPatientPage can be null ???????????

Comment (sherrif10) [reply src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/FindPatientSteps.java:42]: Am cleaning everything @mozzy11 . Thanks for the review

File: src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/FindPatientSteps.java [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+		homePage = loginPage.goToHomePage();
+		
+	}
+	
+	@After(RunTest.HOOK.FINDPATIENT)
+	public void destroy() {
+		quit();
+	}
+	
+	@Given("User clicks on Find Patient App")
```
Comment (mozzy11) [inline src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/FindPatientSteps.java:40]: I think the Right backgound condition should be ,
`Given user Vists home page and clicks clicks on Find Patient App`

Comment (sherrif10) [reply src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/FindPatientSteps.java:40]: > Given user Vists home page and clicks clicks on Find Patient App

Thanks @mozzy11 , these are two different scenarios, its as if you are combining the two  , One  **Userr visits HomePage**  and **user click on FindPatientApp**. In this case we are only using one single function HomePage.goToFindPatientApp.

File: src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/FindPatientSteps.java [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+	public void visitFindPatientPage() {
+		if (findPatientPage != null) {
+			findPatientPage = (FindPatientPage) homePage.goToFindPatientRecord().waitForPage();
```
Comment (mozzy11) [inline src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/FindPatientSteps.java:43]: an also , you methods should reflect he step definition ie 
homePage.clickFindPatientRecordApp()

File: src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/FindPatientSteps.java (hunk 9 above)
Comment (mozzy11) [reply src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/FindPatientSteps.java:40]: Thats right @sherrif10

Comment (mozzy11) [reply src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/FindPatientSteps.java:40]: but make the method more readable

File: src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/FindPatientSteps.java (hunk 8 above)
Comment (sherrif10) [reply src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/FindPatientSteps.java]: Resolved thanks

**Comment by sherrif10** [comment]
Resolved Thanks @mozzy. Looking forward for your reviews

File: src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/page/LoginPage.java (hunk 7 above)
Comment (sherrif10) [reply src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/page/LoginPage.java]: Removed this util .Thanks

File: src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/page/LoginPage.java (hunk 6 above)
Comment (sherrif10) [reply src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/page/LoginPage.java]: Removed this util. thanks

File: src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/test/TestBase.java [hunk 11]
```diff
@@ -47,6 +49,10 @@
 	
 	public static final int MAX_SERVER_STARTUP_IN_MILLISECONDS = 10 * 60 * 1000;
 	
+	protected String firstPatientIdentifier;
```
Comment (mozzy11) [inline src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/test/TestBase.java:52]: hello @sherrif10 , can you please put this variable where its supposed to be??

File: src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/page/FindPatientPage.java [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+	}
+
+	/**
+	 * Finds first record from the result table
+	 * 
+	 * @return patient id
+	 */
+	public String getFirstPatientIdentifier() {
+		getFirstPatientName();
+		return findElement(PATIENT_ID_SEARCH_RESULT).getText();
```
Comment (mozzy11) [inline src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/page/FindPatientPage.java:36]: we already have  an [method](https://github.com/IsantePlus/isanteplus-qaframework/blob/main/src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/page/Page.java#L201) in the parent class that does this same thing

File: src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/page/FindPatientPage.java [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+	}
+
+	public String getFirstPatientName() {
+		return findElement(PATIENT_NAME_SEARCH_RESULT).getText();
```
Comment (mozzy11) [inline src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/page/FindPatientPage.java:40]: same here , there an existing method that already gets Text

File: src/test/java/org/openmrs/contrib/isanteplus/qaframework/RunTest.java [hunk 14]
```diff
@@ -17,5 +18,6 @@
 		public static final String LOGIN = "@login";
                 
                 public static final String REGISTRATION = "@register";
+                public static final String FINDPATIENT = "@findPatient";
```
Comment (mozzy11) [inline src/test/java/org/openmrs/contrib/isanteplus/qaframework/RunTest.java:21]: @sherrif10   , can you also document this filter tag on the ream me ??

**Comment by mozzy11** [comment]
@sherrif10 this generally looks good, can you just clean up a few things in my comments??

File: src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/test/TestBase.java (hunk 11 above)
Comment (sherrif10) [reply src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/test/TestBase.java:52]: This variable is supposed to be in TestBase class to be generic.

File: src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/page/FindPatientPage.java (hunk 12 above)
Comment (sherrif10) [reply src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/page/FindPatientPage.java:36]: Hello @mozzy11 , the method that exists gets text  of an element , but without specifying which element, so I think passing through the exact element its supposed to pick is the right way. is there any other option of getting a **searched patient result**

File: src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/test/TestBase.java (hunk 11 above)
Comment (mozzy11) [reply src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/test/TestBase.java:52]: what do you mean to be generic ??

Comment (sherrif10) [reply src/test/java/org/openmrs/contrib/isanteplus/qaframework/automation/test/TestBase.java:52]: > what do you mean to be generic ??

I mean it method can be reusable

**Comment by mozzy11** [comment]
@sherrif10 , can you also ensure you follow the steps as decribed [here  ](https://paper.dropbox.com/doc/iSantePlus-System-Testing-Protocol-ZcIECdRoBfKswXOFKnWqS)

**Comment by sherrif10** [comment]
> can you also ensure you follow the steps as decribed [here](https://paper.dropbox.com/doc/iSantePlus-System-Testing-Protocol-ZcIECdRoBfKswXOFKnWqS)

Yes @mozzy11  the steps implemented in here are the exact step in the spreadsheet. Thanks

File: src/features/isanteplus/findPatient.feature [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+
+  Background:
+    Given User logs in the system
+
+  @findPatient
+  Scenario: Searching for a patient
+  And User clicks on Find Patient App
+  And User enters missing patient
+  Then Search Page returns no patients
+  And  User enters moses mutesa
```
Comment (mozzy11) [inline src/features/isanteplus/findPatient.feature:11]: you dont need to hard code the name 
can you use the cucmber example tables to load the necessary data??

File: src/features/isanteplus/findPatient.feature [hunk 16]
```diff
@@ -0,0 +1,12 @@
...
+  Background:
+    Given User logs in the system
+
+  @findPatient
+  Scenario: Searching for a patient
+  And User clicks on Find Patient App
+  And  User enters moses mutesa
+  Then Search Page returns patients
+  And  User clicks on first patient
+  Then System loads patient dashboard
```
Comment (mozzy11) [inline src/features/isanteplus/findPatient.feature:12]: can you use    ` Selected patient’s `Cover Page` will be displayed.`

Comment (sherrif10) [reply src/features/isanteplus/findPatient.feature]: Resolved thanks

File: src/features/isanteplus/findPatient.feature (hunk 15 above)
Comment (sherrif10) [reply src/features/isanteplus/findPatient.feature]: Resolved thanks

**Comment by sherrif10** [comment]
Hey @mozzy11  is this ready to be merged. i have addressed all your comments.Thanks alot

File: src/features/isanteplus/findPatient.feature [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+
+  @findPatient
+  Scenario: Searching for a patient
+  And From the home page click ‘rechercher dossier de patient
+  And  Enter First Name or Last Name in “Patient Search” box "<patientSearch>" patientSearch
+  Then Identify patient in list
+  And  Click row with the patient you are searching for
+  Then Selected patient’s “Cover Page” will be displayed
+   Examples:
+      | patientSearch  |
```
[the remark this answers is not in the document]
Comment (sherrif10) [reply src/features/isanteplus/findPatient.feature]: Resolved thanks

**Comment by mozzy11** [comment]
Thanks @sherrif10  for the great work

**Pull request merged**